lib/toolshed/net.ex

```diff
@@ -185,23 +196,130 @@ defmodule Toolshed.Net do
     end
   end
 
-  defp try_connect(address, port, connect_options) do
-    start = System.monotonic_time(:microsecond)
+  defp try_connect(address, connect_options, sequence_number) do
+    interface_address = Keyword.get(connect_options, :ip, :any)
+    timeout = Keyword.get(connect_options, :timeout, 10_000)
+
+    {family, protocol, request_packet_type, reply_packet_type} =
+      case tuple_size(address) do
+        4 -> {:inet, :icmp, 0x08, 0x00}
+        8 -> {:inet6, {:raw, 58}, 0x80, 0x81}
+      end
+
+    packet = <<
+      # Type (Echo request)
+      request_packet_type::size(8),
+      # Code
+      0x00::size(8),
+      # Checksum
+      0x0000::big-integer-size(16),
+      # Identifier
+      0x0001::big-integer-size(16),
+      # Sequence Number
+      sequence_number::big-integer-size(16),
+      # Payload
+      "abcdefghijklmnopqrstuvwxyz012345"::binary
+    >>
+
+    with {:ok, socket} <- :socket.open(family, :dgram, protocol),
+         :ok <- bind(socket, interface_address),
+         start <- System.monotonic_time(:microsecond),
+         :ok <- :socket.sendto(socket, packet, %{family: family, port: 0, addr: address}),
+         {:ok, {_, reply_bytes}} <- :socket.recvfrom(socket, [], timeout),
+         :ok <- check_ping_reply_byte_size(reply_bytes),
+         :ok <- check_ping_reply_type(reply_bytes, reply_packet_type),
+         {:ok, reply} <- parse_ping_reply(reply_bytes, reply_packet_type),
+         :ok <- check_ping_reply_sequence_number(reply, sequence_number) do
+      elapsed_time = System.monotonic_time(:microsecond) - start
+      {:ok, elapsed_time}
+    else
+      error -> error
+    end
+  end
+
+  # OTP 24 changed the return value from `:socket.bind`. This needs to be
+  # accounted for at compile time or else dialyzer will raise an error for
+  # the branch that will never match the code required for backwards
+  # compatibility.
+  if @otp_version >= 24 do
```

Comment:
Another way I could do this is by adding a dialyzer ignore file to ignore the error this creates.